src/System.Private.CoreLib/shared/System/Text/ASCIIEncoding.cs

```diff
@@ -357,564 +398,511 @@ public override unsafe string GetString(byte[] bytes, int byteIndex, int byteCou
+                        goto DestinationTooSmall;
                     }
-                    else
-                        fallbackBuffer.MovePrevious();
 
-                    // Are we throwing or using buffer?
-                    ThrowBytesOverflow(encoder, bytes == byteStart);    // throw?
-                    break;                                              // don't throw, stop
+                    pinnedBytes = pinnedBytes.Slice(bytesWrittenThisReplacement);
                 }
 
-                // Go ahead and add it
-                *bytes = unchecked((byte)ch);
-                bytes++;
+                pinnedChars = pinnedChars.Slice(1); // in either case, we only consumed a single char
+
+                if (!pinnedChars.IsEmpty)
+                {
+                    // If more input data remains, count the number of ASCII characters.
+
+                    numElementsConvertedJustNow = (int)AsciiUtility.NarrowUtf16ToAscii(
+                        pUtf16Buffer: (char*)Unsafe.AsPointer(ref MemoryMarshal.GetReference(pinnedChars)),
+                        pAsciiBuffer: (byte*)Unsafe.AsPointer(ref MemoryMarshal.GetReference(pinnedBytes)),
+                        elementCount: (uint)Math.Min(pinnedChars.Length, pinnedBytes.Length));
+
+                    Debug.Assert(0 <= numElementsConvertedJustNow && numElementsConvertedJustNow <= Math.Min(pinnedChars.Length, pinnedBytes.Length));
+
+                    // Skip over all the ASCII data we just converted.
+
+                    pinnedChars = pinnedChars.Slice(numElementsConvertedJustNow);
+                    pinnedBytes = pinnedBytes.Slice(numElementsConvertedJustNow);
+                }
             }
 
-            // Need to do encoder stuff
+        Finish:
+
+            // If an encoder is available, signal to the encoder how many chars we were able to consume.
+
             if (encoder != null)
             {
-                // Fallback stuck it in encoder if necessary, but we have to clear MustFlush cases
-                if (fallbackBuffer != null && !fallbackBuffer.bUsedEncoder)
-                    // Clear it in case of MustFlush
-                    encoder._charLeftOver = (char)0;
-
-                // Set our chars used count
-                encoder._charsUsed = (int)(chars - charStart);
+                encoder._charsUsed = originalCharCount - pinnedChars.Length;
             }
 
-            Debug.Assert(fallbackBuffer == null || fallbackBuffer.Remaining == 0 ||
-                (encoder != null && !encoder._throwOnOverflow),
-                "[ASCIIEncoding.GetBytes]Expected Empty fallback buffer at end");
+            return originalByteCount - pinnedBytes.Length; // total bytes written
 
-            return (int)(bytes - byteStart);
+        DestinationTooSmall:
+
+            // Fail if we weren't able to consume any chars at all.
+
+            ThrowBytesOverflow(encoder, nothingEncoded: (originalCharCount == pinnedChars.Length));
```

Comment:
Could this just throw at original place instead of 2 gotos

Reply:
It doesn't always throw. In the event it doesn't throw, we want control to transfer to the end of the method. The reason it's not done inline is that we want forward branches to be predicted not taken.